File: xabber/src/main/java/com/xabber/android/data/extension/rrr/RetractMessageIQ.java
package com.xabber.android.data.extension.rrr;

import org.jivesoftware.smack.packet.IQ;

public class RetractMessageIQ extends IQ {

    public static final String NAMESPACE = RrrManager.NAMESPACE;
    public static final String ELEMENT = "retract-message";
    public static final String SYMMETRIC_ATTRIBUTE = "symmetric";
    public static final String ID_ATTRIBUTE = "id";
    public static final String BY_ATTRIBUTE = "by";

    private String by;
    private String id;
    private boolean symmetric;

    public String getNamespace(){ return NAMESPACE; }
    public String getElementName(){ return ELEMENT; }

    RetractMessageIQ(String by, String id, boolean symmetric){
        super(ELEMENT, NAMESPACE);
        this.setType(Type.set);
        this.by = by;
        this.id = id;
        this.symmetric = symmetric;
        this.setFrom(this.by);
    }

    @Override
    protected IQChildElementXmlStringBuilder getIQChildElementBuilder(IQChildElementXmlStringBuilder xml) {
        xml.attribute(SYMMETRIC_ATTRIBUTE, symmetric);
        xml.attribute(BY_ATTRIBUTE, by);
        xml.attribute(ID_ATTRIBUTE, id);
        xml.rightAngleBracket();
        return xml;
    }
}

File: xabber/src/main/java/com/xabber/xmpp/retract/outgoing/RetractMessageIq.kt
package com.xabber.xmpp.retract.outgoing

import com.xabber.android.data.entity.AccountJid
import com.xabber.android.data.entity.ContactJid

class RetractMessageIq(
    private val messageStanzaId: String,
    private val accountJid: AccountJid,
    private val symmetrically: Boolean = false,
    archiveAddress: ContactJid? = null,
) : AbstractRetractIq(archiveAddress, ELEMENT) {

    init {
        type = Type.set
    }

    override fun getIQChildElementBuilder(xml: IQChildElementXmlStringBuilder) = xml.apply {
        attribute(SYMMETRIC_ATTRIBUTE, symmetrically)
        attribute(BY_ATTRIBUTE, accountJid.bareJid.toString())
        attribute(ID_ATTRIBUTE, messageStanzaId)
        rightAngleBracket()
    }

    private companion object {
        const val ELEMENT = "retract-message"
        const val SYMMETRIC_ATTRIBUTE = "symmetric"
        const val BY_ATTRIBUTE = "by"
        const val ID_ATTRIBUTE = "id"
    }

}